### DISCUSSION TITLE
Docker Swarm introduces problems

### DISCUSSION THREAD

**Discussion by 0x5ECF4ULT** [simple-login/app#560] [Self-hosting question]
Hi!
I've set up SL for my homelab and ran into a weird problem.
**TL;DR:** SL is running on a 4-node Raspi cluster (4B 4GB) in Swarm mode. It feels like only every 5th HTTP request to the frontend produces a response. The stack is running behind Traefik instead of NGINX.

## What happened?
Well... Nothing really. Everything is just so. painfully. slow. I've built a [custom](https://gist.github.com/0x5ECF4ULT/5df0e29c99384c216b8f941bdd514b77) compose file and fired it up. That's about as far as I've got until now. I don't feel like there's an issue with the stack file or my cluster.

## Details?!
Yeah yeah okay I get it. So... the juicy details: stack is up and fully booted. I open up my browser, type in the assigned subdomain (`mail.backbone.localdomain`) and whack `Enter`. Loading... Loading... Very nice. This takes forever and suddenly: `Gateway Timeout`. Huh? I check the stack status and everything's green. Weird... This only happens when there's a configuration issue in Traefik but this one's also good. Okay... so `Ctrl+F5` it is... `Gateway Timeout`<sup>AAAAAAAAAAAAA</sup>. Alright, let's take a look at the logs.
Okay after a quick checkup everything seems fine. The logs don't show anything suspicious or something that could be a runtime failure (yes, I checked **all** the logs). Just a line that tells me that the browser should've got a 302 response. It's just... it doesn't look like it ever came around.
It must be a network issue then, I think to myself.

## Networking issue? Your fault!
Nonono don't get me wrong! I don't feel like there's an issue on my end. I feel like there's something wrong with how the SL webapp handles HTTP requests. Other apps (running on the same Raspi btw) don't have any problems with returning responses and run butter-smooth.

## Fine. What have you done until now?
Tbh I didn't do much troubleshooting. One obvious thing I tried, was to move everything onto the master node and pray that it does work. **\*big red button meeeeeep\*** Nope, still the same issue. Some other things I'd want to try:

- [ ] Use NGINX
- [ ] Run with `docker-compose`
- [ ] ~Yeet everything out of the window~
- [x] Stare at my stack file for hours and pray to find the error
- [x] ~Give up~ (Tried... didn't work)

**Comment by nguyenkims** [comment]
Do you see the request in `sl-app` log? It seems that the request might never reach the `sl-app` container. This container listens on the port 7777 
I don't know much about docker swarm and traefik unfortunately so I hope someone can chime in and help spot the issue.

**Reply by 0x5ECF4ULT** [reply]
I see the request in the `sl-app` container so there's no problem with routing HTTP. As an additional troubleshooting step, I might fire up Wireshark and see if the request finds it's way back.

Here's a quick primer on some major differences: Swarm is an orchestrator like k8s. Once a stack is deployed it can be reached from every node. Communication between the nodes and containers spread over the cluster happens with the help of an overlay network. Basically a virtual network that every node is automatically joined into. This is also called the `ingress` network. On a higher level you could say that all nodes are one massive.
Traefik is a reverse proxy on steroids. It can either run in standalone or Swarm mode. One notable thing about this proxy is probably how it routes packets. You assign labels to a service and on deployment, Traefik will read those and route incoming packets accordingly. This is called dynamic configuration.

**Reply by nguyenkims** [reply]
Can you paste the logs on `sl-app` here? Maybe the timeout comes from a connection problem between `sl-app` and Postfix which is used for sending emails.

Btw thanks for the info about Docker swarm and Traefik!

**Reply by 0x5ECF4ULT** [reply]
Yeah sure
```
[2021-08-17 10:37:52 +0000] [1] [INFO] Starting gunicorn 20.0.4,
>>> URL: http://mail.backbone.localdomain,
MAX_NB_EMAIL_FREE_PLAN is not set, use 5 as default value,
load config file /env/.env,
[2021-08-17 10:37:52 +0000] [10] [INFO] Booting worker with pid: 10,
[2021-08-17 10:37:52 +0000] [9] [INFO] Booting worker with pid: 9,
[2021-08-17 10:37:52 +0000] [1] [INFO] Using worker: sync,
[2021-08-17 10:37:52 +0000] [1] [INFO] Listening at: http://0.0.0.0:7777 (1),
2021-08-17 10:37:57,974 - SL - DEBUG - "/code/app/utils.py:11" - <module>() -  - load words file: /code/local_data/words_alpha.txt,
>>> init logging <<<,
Upload files to local dir,
Paddle param not set,
2021-08-17 10:37:57,986 - SL - DEBUG - "/code/app/utils.py:11" - <module>() -  - load words file: /code/local_data/words_alpha.txt,
2021-08-17 12:25:48,450 - SL - DEBUG - "/code/server.py:444" - after_request() -  - 10.0.0.2 GET / ImmutableMultiDict([]) 302,
2021-08-17 12:26:18,655 - SL - DEBUG - "/code/server.py:444" - after_request() -  - 10.0.0.2 GET /favicon.ico ImmutableMultiDict([]) 302,
2021-08-17 12:26:22,048 - SL - DEBUG - "/code/server.py:444" - after_request() -  - 10.0.0.2 GET /auth/login ImmutableMultiDict([]) 200
```
Sometimes `Ctrl+F5`-ing helps to get me until `/auth/login`

**Reply by nguyenkims** [reply]
I don't see anything abnormal in the log here. Do you have any log after trying to log in? I wonder if POST requests don't somehow reach the web container.

**Reply by 0x5ECF4ULT** [reply]
Me neither. I never came so far to be able to login. POST requests behave just like GET requests. Sometimes they reach the container, sometimes they don't.
